### PR TITLE
Add rpc call `server_definitions` to rippled.

### PR DESCRIPTION
**Pull request by RichardAH**
## High Level Overview of Change
Add a new RPC / ws call: `server_definitions`, which returns an *xrpljs* compatible definitions JSON generated from the rippled instance currently running. This will alleviate definitions juggling, especially with side chains which may have exotic features and serialized fields. Wallet software will be able to query `server_definitions` against a node from the network they want to connect to and immediately know how to speak that node's language, even if new features are added to it in the future.

Example:
```js
> {"command": "server_definitions"}
< {
    "result": {
        "FIELDS": [
            [
                "Generic",
                {
                    "isSerialized": false,
                    "isSigningField": false,
                    "isVLEncoded": false,
                    "nth": 0,
                    "type": "Unknown"
                }
            ],
            [
                "Invalid",
                {
                    "isSerialized": false,
                    "isSigningField": false,
                    "isVLEncoded": false,
                    "nth": -1,
                    "type": "Unknown"
                }
            ],
            [
                "ObjectEndMarker",
                {
                    "isSerialized": false,
                    "isSigningField": true,
                    "isVLEncoded": false,
                    "nth": 1,
                    "type": "STObject"
                }
            ],
        ...
```

### CODE REVIEW

**Comment by sublimator** [comment]
wow, FINALLY

**Comment by intelliot** [comment]
@JST5000 @gregtatcam could you have a look at this?

**Comment by mvadari** [comment]
This looks pretty reasonable to me (it's very similar to the JS version, https://github.com/RichardAH/xrpl-codec-gen), but I'm no C++ expert

**Comment by JST5000** [comment]
> @JST5000 @gregtatcam could you have a look at this?

I can review on Wednesday :)

**Comment by gregtatcam** [comment]
It doesn't build on OSX M1 Ventura 13.3.1, apple clang 14.0.0, cmake 3.23.2. Errors:
```
In file included from /.../rippled/src/ripple/rpc/handlers/ServerInfo.cpp:34:
/.../rippled/src/magic/magic_enum.h:590:18: error: constexpr variable 'valid' must be initialized by a constant expression
  constexpr bool valid[sizeof...(I)] = {is_valid<E, ualue<E, Min, IsFlags>(I)>()...};
                 ^                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/.../rippled/src/magic/magic_enum.h:616:10: note: in instantiation of function template specialization 'magic_enum::detail::values<ripple::SerializedTypeID, false, -20000, 0UL, 1UL, 2UL, 3UL, 4UL, 5UL, 6UL, 7UL, 8UL, 9UL, 10UL, 11UL, 12UL, 13UL, 14UL, 15UL, 16UL, 17UL, 18UL, 19UL,...,40000UL>' requested here
  return values<E, IsFlags, min>(std::make_index_sequence<range_size>{});

/.../rippled/src/magic/magic_enum.h:632:37: note: in instantiation of function template specialization 'magic_enum::detail::values<ripple::SerializedTypeID, false, int>' requested here
    constexpr auto default_values = values<E, false>();
                                    ^
/.../rippled/src/magic/magic_enum.h:648:36: note: in instantiation of function template specialization 'magic_enum::detail::is_flags_enum<ripple::SerializedTypeID, int>' requested here
inline constexpr bool is_flags_v = is_flags_enum<E>();
                                   ^
/.../rippled/src/magic/magic_enum.h:651:50: note: in instantiation of variable template specialization 'magic_enum::detail::is_flags_v<ripple::SerializedTypeID>' requested here
inline constexpr std::array values_v = values<E, is_flags_v<E>>();
                                                 ^
/.../rippled/src/magic/magic_enum.h:657:33: note: in instantiation of variable template specialization 'magic_enum::detail::values_v' requested here
inline constexpr auto count_v = values_v<E>.size();
                                ^
/.../rippled/src/magic/magic_enum.h:686:77: note: in instantiation of variable template specialization 'magic_enum::detail::count_v' requested here
inline constexpr std::array entries_v = entries<E>(std::make_index_sequence<count_v<E>>{});
                                                                            ^
/.../rippled/src/magic/magic_enum.h:689:29: note: in instantiation of variable template specialization 'magic_enum::detail::entries_v' requested here
using entries_t = decltype((entries_v<D>));
                            ^
/.../rippled/src/magic/magic_enum.h:1235:25: note: in instantiation of template type alias 'entries_t' requested here
[[nodiscard]] constexpr auto enum_entries() noexcept -> detail::enable_if_t<E, detail::entries_t<E>> {
                        ^
/.../rippled/src/ripple/rpc/handlers/ServerInfo.cpp:150:35: note: while substituting explicitly-specified template arguments into function template 'enum_entries' 
        for (auto [value, name] : magic_enum::enum_entries<SerializedTypeID>())
                                  ^
/.../rippled/src/magic/magic_enum.h:281:3: note: constexpr evaluation hit maximum step limit; possible infinite loop?
  return {}; // Invalid name.
  ^
/.../rippled/src/magic/magic_enum.h:453:27: note: in call to 'pretty_name({&__PRETTY_FUNCTION__[91], 4})'
    constexpr auto name = pretty_name({__PRETTY_FUNCTION__, sizeof(__PRETTY_FUNCTION__) - 2});
                          ^
/.../rippled/src/magic/magic_enum.h:505:12: note: in call to 'n()'
    return n<E, v>().size() != 0;
           ^
/.../rippled/src/magic/magic_enum.h:590:41: note: in call to 'is_valid()'
  constexpr bool valid[sizeof...(I)] = {is_valid<E, ualue<E, Min, IsFlags>(I)>()...};
                                        ^
/.../rippled/src/magic/magic_enum.h:591:25: error: constexpr variable 'count' must be initialized by a constant expression
  constexpr std::size_t count = values_count(valid);
                        ^       ~~~~~~~~~~~~~~~~~~~
/.../rippled/src/magic/magic_enum.h:579:9: note: initializer of 'valid' is not a constant expression
    if (valid[i]) {
        ^
/.../rippled/src/magic/magic_enum.h:591:33: note: in call to 'values_count(valid)'
  constexpr std::size_t count = values_count(valid);
                                ^
/.../rippled/src/magic/magic_enum.h:590:18: note: declared here
  constexpr bool valid[sizeof...(I)] = {is_valid<E, ualue<E, Min, IsFlags>(I)>()...};
                 ^
/.../rippled/src/magic/magic_enum.h:593:17: error: constexpr if condition is not a constant expression
  if constexpr (count > 0) {
                ^~~~~~~~~
/.../rippled/src/magic/magic_enum.h:593:17: note: initializer of 'count' is not a constant expression
/.../rippled/src/magic/magic_enum.h:591:25: note: declared here
  constexpr std::size_t count = values_count(valid);
                        ^
/.../rippled/src/magic/magic_enum.h:594:14: error: variable-sized object may not be initialized
    E values[count] = {};
             ^~~~~
/.../rippled/src/magic/magic_enum.h:601:54: error: non-type template argument is not a constant expression
    return to_array(values, std::make_index_sequence<count>{});
                                                     ^~~~~
/.../rippled/src/magic/magic_enum.h:601:54: note: initializer of 'count' is not a constant expression
/.../rippled/src/magic/magic_enum.h:591:25: note: declared here
  constexpr std::size_t count = values_count(valid);
                        ^
/.../rippled/src/magic/magic_enum.h:648:36: error: constexpr variable 'is_flags_v<ripple::SerializedTypeID>' must be initialized by a constant expression
inline constexpr bool is_flags_v = is_flags_enum<E>();
                                   ^~~~~~~~~~~~~~~~~~
/.../rippled/src/magic/magic_enum.h:651:50: note: in instantiation of variable template specialization 'magic_enum::detail::is_flags_v<ripple::SerializedTypeID>' requested here
inline constexpr std::array values_v = values<E, is_flags_v<E>>();
                                                 ^
/.../rippled/src/magic/magic_enum.h:657:33: note: in instantiation of variable template specialization 'magic_enum::detail::values_v' requested here
inline constexpr auto count_v = values_v<E>.size();
                                ^
/.../rippled/src/magic/magic_enum.h:686:77: note: in instantiation of variable template specialization 'magic_enum::detail::count_v' requested here
inline constexpr std::array entries_v = entries<E>(std::make_index_sequence<count_v<E>>{});
                                                                            ^
/.../rippled/src/magic/magic_enum.h:689:29: note: in instantiation of variable template specialization 'magic_enum::detail::entries_v' requested here
using entries_t = decltype((entries_v<D>));
                            ^
/.../rippled/src/magic/magic_enum.h:1235:25: note: in instantiation of template type alias 'entries_t' requested here
[[nodiscard]] constexpr auto enum_entries() noexcept -> detail::enable_if_t<E, detail::entries_t<E>> {
                        ^
/.../rippled/src/ripple/rpc/handlers/ServerInfo.cpp:150:35: note: while substituting explicitly-specified template arguments into function template 'enum_entries' 
        for (auto [value, name] : magic_enum::enum_entries<SerializedTypeID>())
                                  ^
In file included from /.../rippled/src/ripple/rpc/handlers/ServerInfo.cpp:34:
/.../rippled/src/magic/magic_enum.h:651:40: error: no matching function for call to 'values'
inline constexpr std::array values_v = values<E, is_flags_v<E>>();
                                       ^~~~~~~~~~~~~~~~~~~~~~~~
/.../rippled/src/magic/magic_enum.h:657:33: note: in instantiation of variable template specialization 'magic_enum::detail::values_v' requested here
inline constexpr auto count_v = values_v<E>.size();
                                ^
/.../rippled/src/magic/magic_enum.h:686:77: note: in instantiation of variable template specialization 'magic_enum::detail::count_v' requested here
inline constexpr std::array entries_v = entries<E>(std::make_index_sequence<count_v<E>>{});
                                                                            ^
/.../rippled/src/magic/magic_enum.h:689:29: note: in instantiation of variable template specialization 'magic_enum::detail::entries_v' requested here
using entries_t = decltype((entries_v<D>));
                            ^
/.../rippled/src/magic/magic_enum.h:1235:25: note: in instantiation of template type alias 'entries_t' requested here
[[nodiscard]] constexpr auto enum_entries() noexcept -> detail::enable_if_t<E, detail::entries_t<E>> {
                        ^
/.../rippled/src/ripple/rpc/handlers/ServerInfo.cpp:150:35: note: while substituting explicitly-specified template arguments into function template 'enum_entries' 
        for (auto [value, name] : magic_enum::enum_entries<SerializedTypeID>())
                                  ^
/.../rippled/src/magic/magic_enum.h:608:16: note: candidate template ignored: invalid explicitly-specified argument for template parameter 'IsFlags'
constexpr auto values() noexcept {
               ^
/.../rippled/src/magic/magic_enum.h:590:16: note: candidate function template not viable: requires 1 argument, but 0 were provided
constexpr auto values(std::index_sequence<I...>) noexcept {
               ^
/.../rippled/src/ripple/rpc/handlers/ServerInfo.cpp:150:35: error: no matching function for call to 'enum_entries'
        for (auto [value, name] : magic_enum::enum_entries<SerializedTypeID>())
                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/.../rippled/src/magic/magic_enum.h:1235:30: note: candidate template ignored: substitution failure [with E = ripple::SerializedTypeID]
[[nodiscard]] constexpr auto enum_entries() noexcept -> detail::enable_if_t<E, detail::entries_t<E>> {
                             ^
/.../rippled/src/ripple/rpc/handlers/ServerInfo.cpp:153:36: error: member reference base type 'void' is not a structure or union
                translate(STR(name).substr(4) /* remove STI_ */);
                          ~~~~~~~~~^~~~~~~
/.../rippled/src/ripple/rpc/handlers/ServerInfo.cpp:154:34: error: no matching function for call to 'stoi'
            int32_t type_value = std::stoi(STR(value));
```

**Comment by gregtatcam** [comment]
It doesn't build on Windows 11, VS 2022. Errors:
```
 C:\Users\rippled\src\magic\magic_enum.h(593,23): error C2131: expression did not evaluate to a constant (compiling source file C:                    \Users\rippled\src\ripple\rpc\handlers\ServerInfo.cpp) 
   C:\Users\rippled\src\magic/magic_enum.h(272,32): message : failure was caused by evaluation exceeding step limit of 1048576 (/constexpr:           steps<NUMBER>) (compiling source file C:\Users\rippled\src\ripple\rpc\handlers\ServerInfo.cpp) [C:\Users\rippled\out\build\x64-Release\rippled.      vcxproj]
   C:\Users\rippled\src\magic/magic_enum.h(608): message : see reference to function template instantiation 'auto magic_enum::detail::values<ripple:: SerializedTypeID,false,-20000,0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,...,39961,39962,39963,39964,39965,39966,  39967,39968,39969,39970,39971,39972,39973,39974,39975,39976,39977,39978,39979,39980,39981,39982,39983,39984,39985,39986,39987,39988,39989,39990,     39991,39992,39993,39994,39995,39996,39997,39998,39999,40000>) noexcept' being compiled (compiling source file C:                                     \Users\rippled\src\ripple\rpc\handlers\ServerInfo.cpp) [C:\Users\rippled\out\build\x64-Release\rippled.vcxproj]
```

**Review by gregtatcam — CHANGES_REQUESTED** [review]

File: src/ripple/rpc/handlers/ServerInfo.cpp [hunk 1]
```diff
@@ -21,14 +21,377 @@
...
+private:
+    Json::Value
+    generate()
+    {
+        // RH TODO: probably a better way to do this?
+#define STR(x)                      \
+    ([&] {                          \
+        std::ostringstream ss;      \
+        return ss << (x), ss.str(); \
+    }())
```
Comment (gregtatcam) [inline src/ripple/rpc/handlers/ServerInfo.cpp:67]: Consider using a function instead. It'll fix compile error with apple `clang` (see comments below). Something like this should work.
```
template <typename T>
std::string
str(T const& t)
{
    std::ostringstream ss;
    ss << t;
    return ss.str();
}
```

**Review by gregtatcam — CHANGES_REQUESTED** [review]

**Comment by gregtatcam** [comment]
There are differences between `xrpl-codec-gen` (js) and `server_definitions` (rpc) output:

Different values or missing keys:

(key:)key1                            key2                        js value    rpc value
------------------------    ----                         --------   ----------
LEDGER_ENTRY_TYPES    Child                       -2             (missing)
TRANSACTION_RESULTS  tecHOOK_ERROR (missing) 154
LEDGER_ENTRY_TYPES    Any                         -3             0
FIELDS:ObjectEndMarker  IsSerialized            true          false
FIELDS:ArrayEndMarker    isSerialized            true          false
FIELDS:LedgerEntry           isSerialized            false        true
FIELDS:LedgerEntry           isSigningField        true         false
FIELDS:Transaction            isSerialized            false        true
FIELDS:Transaction            isSigningField        true         false
FIELDS:Validation               isSerialized            false        true
FIELDS:Validation               isSigningField        true         false
FIELDS:Metadata                isSigningField        true         false

Different keys:

key1                                     js-key2                        rpc-key2
------------------------    ----                             --------
TRANSACTION_TYPES      TicketCancel              SpinalTap
TRANSACTION_TYPES      NickNameSet             NicknameSet
TRANSACTION_TYPES      EnableAmendement  Amendement
TRANSACTION_TYPES      SetFee                        Fee

**Review by HowardHinnant — CHANGES_REQUESTED** [review]
Needs rebase to develop and ensure compiles.

**Comment by RichardAH** [comment]
Firstly thanks for the review. Before I run through and address these, the compilation issue with/on certain compilers and libstdc++ combinations is an open issue in magic_enum repo here: https://github.com/Neargye/magic_enum/issues/229

Fwiw it compiles with both sets of tool chains I typically use. If the compilation of magic_enum is a blocker then I will wait until it's resolved. Perhaps one of you who is having trouble building it locally can help the magic_enum project fix the issue?

**Comment by intelliot** [comment]
Looks like this doesn't build on Mac and Windows, which blocks this PR.

Also, is it necessary to introduce a new library? The need for a new dependency needs to be balanced against the need for the feature.

**Comment by seelabs** [comment]
We've been relatively conservative in adding new dependencies to the server; I don't think the "magic enum" library is worth adding as a dependency. I'm not saying I'm against ever adding new dependencies, but in this case it's a fair amount of complexity for a smallish win. I'm also not saying I get to decide what dependencies get added, just that my $0.02 is we should continue to be conservative.

**Comment by seelabs** [comment]
@RichardAH 

I should have articulated why I'm against it better in my previous comment. Internally people have proposed adding libraries for many things: better hash tables, libraries for better support for C++ coroutine support (libunifex), advanced metaprogramming (hanna), as well as others. I'm usually the one who objects and tries to keep the number of dependencies low. I don't want developers to need to learn too many libraries, and I also don't want to rely on libraries that may be abandoned.

For this specific library, this magic enum library is actively developed, easy to use, and fills a hole that C++ should have filled a long time ago (although the implementation is complex and relies on compiler intrinsics rather than standard C++). It wouldn't be the worst thing if we added it and I do see where the benefit it brings to the code. On the other hand, I still think this is a "nice to have" rather than a "must have", and I don't think we want to start adding a large number of "nice to have" libraries. We're going to end up with a union of every developer's favorite libraries.

So while I am against adding this dependency I won't stand in the way if other's are in favor.

As a side note, I'm not worried about build issues - the fact that it's header-only does make building easier, but it wouldn't bother me if it wasn't. I'm also not against adding all new dependencies. At some point I would like to float the idea of adding some larger dependencies.

File: src/ripple/rpc/handlers/ServerInfo.cpp [hunk 2]
```diff
@@ -21,14 +21,377 @@
...
 #include <ripple/rpc/impl/TransactionSign.h>
+#include <boost/algorithm/string.hpp>
+#include <magic/magic_enum.h>
+#include <sstream>
+
+#define MAGIC_ENUM(x)                           \
+    template <>                                 \
+    struct magic_enum::customize::enum_range<x> \
+    {                                           \
+        static constexpr int min = -20000;      \
```
[the remark this answers is not in the document]
Comment (Neargye) [reply src/ripple/rpc/handlers/ServerInfo.cpp:41]: Range is really too big. I will look at possible improvements

**Comment by RichardAH** [comment]
Our revised position is that if `server_definitions` rpc _isn't_ present on the node then the client should just use assume default definitions. Mainnet will continue to need clients to manually update definitions when mainnet wants to add a new feature. If you want this re-opened let me know.

**Comment by Neargye** [comment]
@intelliot  
You can check the [latest changes](https://github.com/Neargye/magic_enum/commit/629f7b09af87a83541eaeabaeefb92cbfa6f4b0e), the problem with build was caused by large ranges in `constexpr bool valid[sizeof...(I)] = { is_valid<E, ualue<E>(I)>()... };`. Changes allows to use big range

**Comment by mvadari** [comment]
I wrote a version of this PR that doesn't use `magic_enum`: https://github.com/XRPLF/rippled/pull/4703

**Pull request closed without merging**